File: src/lib/settingsSearch.ts
/**
 * Settings search helpers: synonyms, fuzzy matching, typeahead suggestions
 * and persistence of the last query.
 */

const QUERY_KEY = 'nutrient-tracker-settings-search';

export function loadSettingsQuery(): string {
  try {
    return localStorage.getItem(QUERY_KEY) ?? '';
  } catch {
    return '';
  }
}

export function saveSettingsQuery(query: string): void {
  try {
    if (query) localStorage.setItem(QUERY_KEY, query);
    else localStorage.removeItem(QUERY_KEY);
  } catch {
    /* storage unavailable – ignore */
  }
}

/** Groups of interchangeable words. Any word matches every other in its group. */
const SYNONYM_GROUPS: string[][] = [
  ['backup', 'restore', 'save', 'archive', 'snapshot'],
  ['export', 'download', 'share', 'save file'],
  ['import', 'upload', 'load', 'merge'],
  ['appearance', 'theme', 'look', 'style', 'design', 'skin', 'texture'],
  ['color', 'colour', 'accent', 'hue', 'palette'],
  ['dark mode', 'night mode', 'dark'],
  ['goals', 'targets', 'limits', 'macros'],
  ['nutrition', 'nutrients', 'vitamins', 'minerals'],
  ['sign out', 'log out', 'logout', 'signout', 'leave'],
  ['account', 'profile', 'user', 'login'],
  ['premium', 'pro', 'paid', 'subscription', 'unlock', 'donation'],
  ['feedback', 'contact', 'support', 'help', 'bug', 'issue', 'report'],
  ['offline', 'no internet', 'airplane', 'local'],
  ['error', 'log', 'crash', 'debug', 'diagnostics'],
  ['apple health', 'healthkit', 'health', 'shortcuts'],
  ['delete', 'remove', 'erase', 'wipe'],
  ['sync', 'cloud', 'icloud', 'server'],
  ['test', 'checklist', 'qa', 'verify'],
  ['food', 'library', 'database', 'foods'],
  ['recipe', 'recipes', 'meal', 'mealplan'],
];

/** Expand a query with its synonyms. */
export function expandQuery(query: string): string[] {
  const q = query.trim().toLowerCase();
  if (!q) return [];
  const out = new Set<string>([q]);
  for (const group of SYNONYM_GROUPS) {
    if (group.some((w) => w.includes(q) || q.includes(w))) {
      group.forEach((w) => out.add(w));
    }
  }
  return [...out];
}

/** Lightweight subsequence fuzzy match ("bckp" -> "backup"). */
export function fuzzyMatch(needle: string, haystack: string): boolean {
  const n = needle.toLowerCase();
  const h = haystack.toLowerCase();
  if (!n) return true;
  if (h.includes(n)) return true;
  if (n.length < 3) return false;
  let i = 0;
  for (const ch of h) {
    if (ch === n[i]) i += 1;
    if (i === n.length) return true;
  }
  return false;
}

/** Does any keyword match the query (direct, synonym, or fuzzy)? */
export function matchesKeywords(query: string, keywords: string[]): boolean {
  const q = query.trim().toLowerCase();
  if (!q) return true;
  const variants = expandQuery(q);
  return keywords.some((k) => variants.some((v) => fuzzyMatch(v, k) || k.toLowerCase().includes(v)));
}

export interface SettingsSuggestion {
  /** Visible label of the setting. */
  label: string;
  /** Section it belongs to (shown as secondary text). */
  section: string;
  /** Query to apply when picked. */
  query: string;
  keywords: string[];
}

export const SETTINGS_INDEX: SettingsSuggestion[] = [
  { label: 'Sign out', section: 'Account', query: 'sign out', keywords: ['sign out', 'log out', 'account', 'logout'] },
  { label: 'Premium status', section: 'Account', query: 'premium', keywords: ['premium', 'unlock', 'pro', 'donation'] },
  { label: 'Delete account', section: 'Account', query: 'delete account', keywords: ['delete account', 'remove account', 'erase'] },
  { label: 'Daily goals', section: 'Goals & Nutrition', query: 'goals', keywords: ['goals', 'daily goals', 'targets', 'calories'] },
  { label: 'Nutrient library', section: 'Goals & Nutrition', query: 'nutrient library', keywords: ['nutrient library', 'json', 'import', 'export', 'foods'] },
  { label: 'Theme & appearance', section: 'Appearance', query: 'appearance', keywords: ['appearance', 'theme', 'dark mode', 'color', 'accent'] },
  { label: 'Theme packs', section: 'Appearance', query: 'texture pack', keywords: ['texture', 'pack', 'theme packs', 'gallery'] },
  { label: 'Apple Health export', section: 'Data & Sync', query: 'apple health', keywords: ['apple health', 'healthkit', 'shortcuts', 'export'] },
  { label: 'Backup & restore', section: 'Advanced', query: 'backup', keywords: ['backup', 'restore', 'archive', 'save'] },
  { label: 'Offline simulation', section: 'Advanced', query: 'offline', keywords: ['offline', 'simulation', 'local'] },
  { label: 'Error log', section: 'Advanced', query: 'error log', keywords: ['error', 'log', 'debug', 'crash'] },
  { label: 'Test checklist', section: 'Advanced', query: 'test checklist', keywords: ['test', 'checklist', 'qa', 'diagnostics'] },
  { label: 'Send feedback', section: 'Support', query: 'feedback', keywords: ['feedback', 'bug', 'feature', 'contact', 'support'] },
];

/** Rank suggestions for the current query. Empty query -> no suggestions. */
export function getSuggestions(query: string, limit = 5): SettingsSuggestion[] {
  const q = query.trim().toLowerCase();
  if (!q) return [];
  const scored = SETTINGS_INDEX.map((item) => {
    const label = item.label.toLowerCase();
    let score = 0;
    if (label.startsWith(q)) score = 100;
    else if (label.includes(q)) score = 80;
    else if (item.keywords.some((k) => k.toLowerCase().includes(q))) score = 60;
    else if (matchesKeywords(q, [item.label, ...item.keywords])) score = 30;
    return { item, score };
  })
    .filter((s) => s.score > 0)
    .sort((a, b) => b.score - a.score);
  return scored.slice(0, limit).map((s) => s.item);
}

File: ios-native/NutriTrack/Search/SettingsSearch.swift
import Foundation

/// Settings search: synonym expansion, subsequence fuzzy matching, ranked
/// typeahead suggestions, and persistence of the last query.
enum SettingsSearch {
    private static let queryKey = "nutritrack.settings.query"

    static var lastQuery: String {
        get { UserDefaults.standard.string(forKey: queryKey) ?? "" }
        set {
            if newValue.isEmpty { UserDefaults.standard.removeObject(forKey: queryKey) }
            else { UserDefaults.standard.set(newValue, forKey: queryKey) }
        }
    }

    /// Any word in a group matches every other word in that group.
    static let synonymGroups: [[String]] = [
        ["backup", "restore", "save", "archive", "snapshot"],
        ["export", "download", "share", "save file"],
        ["import", "upload", "load", "merge"],
        ["appearance", "theme", "look", "style", "design", "skin", "texture"],
        ["color", "colour", "accent", "hue", "palette"],
        ["dark mode", "night mode", "dark"],
        ["goals", "targets", "limits", "macros"],
        ["nutrition", "nutrients", "vitamins", "minerals"],
        ["sign out", "log out", "logout", "leave"],
        ["account", "profile", "user", "login"],
        ["premium", "pro", "paid", "subscription", "unlock", "donation"],
        ["feedback", "contact", "support", "help", "bug", "issue", "report"],
        ["offline", "no internet", "airplane", "local"],
        ["error", "log", "crash", "debug", "diagnostics"],
        ["apple health", "healthkit", "health", "shortcuts"],
        ["delete", "remove", "erase", "wipe"],
        ["sync", "cloud", "icloud", "server"],
        ["food", "library", "database", "foods"],
        ["recipe", "recipes", "meal", "mealplan"],
        ["history", "undo", "activity", "recent"],
    ]

    static func expand(_ query: String) -> [String] {
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !q.isEmpty else { return [] }
        var out: Set<String> = [q]
        for group in synonymGroups where group.contains(where: { $0.contains(q) || q.contains($0) }) {
            group.forEach { out.insert($0) }
        }
        return Array(out)
    }

    /// Substring first, then an in-order subsequence test ("bckp" → "backup").
    static func fuzzyMatch(_ needle: String, _ haystack: String) -> Bool {
        let n = needle.lowercased(), h = haystack.lowercased()
        if n.isEmpty { return true }
        if h.contains(n) { return true }
        guard n.count >= 3 else { return false }
        var i = n.startIndex
        for ch in h where ch == n[i] {
            i = n.index(after: i)
            if i == n.endIndex { return true }
        }
        return false
    }

    static func matches(query: String, keywords: [String]) -> Bool {
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !q.isEmpty else { return true }
        let variants = expand(q)
        return keywords.contains { k in
            variants.contains { fuzzyMatch($0, k) || k.lowercased().contains($0) }
        }
    }

    struct Suggestion: Identifiable, Hashable {
        var id: String { label }
        let label: String
        let section: String
        let query: String
        let keywords: [String]
    }

    static let index: [Suggestion] = [
        .init(label: "Daily goals", section: "Goals & Nutrition", query: "goals",
              keywords: ["goals", "daily goals", "targets", "calories", "protein", "carbs", "fat", "fiber", "water"]),
        .init(label: "Serving size", section: "Goals & Nutrition", query: "serving",
              keywords: ["serving", "portion", "grams", "default"]),
        .init(label: "Per-weekday goals", section: "Goals & Nutrition", query: "weekday",
              keywords: ["weekday", "weekly", "day", "monday", "override"]),
        .init(label: "Nutrient library", section: "Goals & Nutrition", query: "nutrient library",
              keywords: ["nutrient library", "json", "import", "export", "foods"]),
        .init(label: "Theme & appearance", section: "Appearance", query: "appearance",
              keywords: ["appearance", "theme", "dark mode", "color", "accent"]),
        .init(label: "Apple Health export", section: "Data & Sync", query: "apple health",
              keywords: ["apple health", "healthkit", "shortcuts", "export"]),
        .init(label: "Backup & restore", section: "Advanced", query: "backup",
              keywords: ["backup", "restore", "archive", "save", "import", "overwrite"]),
        .init(label: "Settings history & undo", section: "Advanced", query: "history",
              keywords: ["history", "undo", "activity", "recent", "revert"]),
        .init(label: "Delete all data", section: "Advanced", query: "delete",
              keywords: ["delete", "erase", "wipe", "reset"]),
    ]

    /// Label prefix 100, label contains 80, keyword contains 60, fuzzy 30.
    static func suggestions(for query: String, limit: Int = 5) -> [Suggestion] {
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !q.isEmpty else { return [] }
        return index.compactMap { item -> (Suggestion, Int)? in
            let label = item.label.lowercased()
            let score: Int
            if label.hasPrefix(q) { score = 100 }
            else if label.contains(q) { score = 80 }
            else if item.keywords.contains(where: { $0.lowercased().contains(q) }) { score = 60 }
            else if matches(query: q, keywords: [item.label] + item.keywords) { score = 30 }
            else { return nil }
            return (item, score)
        }
        .sorted { $0.1 > $1.1 }
        .prefix(limit)
        .map(\.0)
    }

    /// Highlights every occurrence of `query` inside `text`.
    static func highlighted(_ text: String, query: String) -> AttributedString {
        var attributed = AttributedString(text)
        let q = query.trimmingCharacters(in: .whitespaces)
        guard !q.isEmpty else { return attributed }
        var search = attributed.startIndex..<attributed.endIndex
        while let range = attributed[search].range(of: q, options: .caseInsensitive) {
            attributed[range].backgroundColor = .yellow.opacity(0.35)
            attributed[range].inlinePresentationIntent = .stronglyEmphasized
            guard range.upperBound < attributed.endIndex else { break }
            search = range.upperBound..<attributed.endIndex
        }
        return attributed
    }
}
